datasource/src/main/scala/quasar/physical/s3/S3DataSourceModule.scala

```diff
@@ -43,13 +48,21 @@ object S3DataSourceModule extends LightweightDatasourceModule {
       : F[InitializationError[Json] \/ Disposable[F, Datasource[F, Stream[F, ?], ResourcePath, Stream[F, Data]]]] = {
     config.as[S3Config].result match {
       case Right(s3Config) => {
-        Http1Client[F]() map { client =>
-          val ds: Datasource[F, Stream[F, ?], ResourcePath, Stream[F, Data]] =
-            new S3DataSource[F](client, s3Config)
+        Http1Client[F]() flatMap { client =>
+          val s3Ds = new S3DataSource[F](client, s3Config)
+          val ds: Datasource[F, Stream[F, ?], ResourcePath, Stream[F, Data]] = s3Ds
```

Comment:
I think you can merge `s3Ds` and `ds` together into 1 val?

Reply:
I think we need this so we can upcast `S3DataSource[F]` to `Datasource[F, Stream[F, ?], ResourcePath, Stream[F, Data]]`?

Reply:
Ah right, just adding a type annotation for `s3Ds` would have made that difference more explicit.

Another option, without using `ds`, would be:

```
s3Ds.isLive.ifM({
            val disposable = Disposable((s3Ds: Datasource[F, Stream[F, ?], ResourcePath, Stream[F, Data]]), client.shutdown)
```

But fine with me to leave like this.